### PR TITLE
Bug 1809305 - Allow user to copy an image to the clipboard

### PR DESCRIPTION
**Pull request by t-p-white**
[copy image snackbar.webm](https://user-images.githubusercontent.com/1833156/221835892-6e2192f6-d483-4e42-acf6-5b57f8d68ab0.webm)

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1809305
https://bugzilla.mozilla.org/show_bug.cgi?id=1819746

### CODE REVIEW

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:656]: I welcome any discussion around if this is the best place to store the 'temporary file'.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:678]: There is an interesting 'bug' that is caught here that can be triggered if the device date\time is modified, it can trigger an SSLHandshakeException. I would like to investigate this in a separate PR for this as to not block this feature.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:95]: menu position for 'copy image' conforms to standard seen in other browsers

**Comment by Mugurell** [comment]
This scenario seems very similar to the one for sharing images.
I recommend checking https://github.com/mozilla-mobile/android-components/pull/9420 which has some discussions about how to integrate this functionality in our MVI architecture and I think it would be possible to reuse most of the existing functionality in the `ShareDownloadFeature`, just that instead of sharing based on the `share` state we could trigger copying the image based on a new `copy` state.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (Mugurell) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:82]: nit: Renamings like this or changes that are not needed for the issue at hand are usually avoided since they grow the size of the PR and make the review process harder.

**Comment by t-p-white** [comment]
> This scenario seems very similar to the one for sharing images. I recommend checking [mozilla-mobile/android-components#9420](https://github.com/mozilla-mobile/android-components/pull/9420) which has some discussions about how to integrate this functionality in our MVI architecture and I think it would be possible to reuse most of the existing functionality in the `ShareDownloadFeature`, just that instead of sharing based on the `share` state we could trigger copying the image based on a new `copy` state.

Thanks for highlighting this @Mugurell, I shall take a look 👍 It may be worth holding off reviewing.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [reply android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:82]: Appreciate that, I'm mindful of this 🙂 I felt for this case that there was a small enough number of changes and the comparison would be straightforward. Thinking about it now, a separate commit would have been useful. For these kinds of changes, is it recommended creating a separate issue?

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:570]: Moved SnackbarDelegate and DefaultSnackbarDelegate to SupportUtils

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [reply android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt]: Just FYI, since changing to use the existing feature it does not include this renaming

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt
Comment (t-p-white) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt:66]: This doc has been moved to the classes that implement it

File: android-components/components/support/ktx/build.gradle
Comment (t-p-white) [inline android-components/components/support/ktx/build.gradle:60]: out of scope for this, lines 47-49 are duplicates and need removing. Will create an issue for this

File: android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeatureTest.kt
Comment (t-p-white) [inline android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeatureTest.kt:1]: For reference, this file is similar to ShareDownloadTest

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt
Comment (t-p-white) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt:46]: Dereived from the orginial ShareDownloadFeature

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt
Comment (t-p-white) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt:96]: start & startsharing have moved to ShareDownloadFeature

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt
Comment (Mugurell) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt:28]: This documentation is confusing, same as on `ShareDownloadFeature`.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt
Comment (Mugurell) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt:79]: I see here using the same 1 second timeout.
This was needed for sharing because we wanted continuitym, it was important to show the share options for the image after at most 1 second of the user tapping "Share image".
In the case of copying an image to the clipboard I'd say that this timeout can be made bigger and would also think of this feature as a backup for when sharing could not complete in one second - users could still copy the image and until going to where they want to paste it we'd have more chances to download and add it to the clipboard.
Not really a blocker, wanted to just raise this as something to think about.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt
Comment (Mugurell) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt:29]: Saw that these lines exist in the current implementation (thinking that they remained from a previous implementation based on middlewares).
They are confusing and should be removed.
This is not a `Middleware` but a `LifecycleAwareFeature` which does not intercept actions but rather reacts to specific store updates.

**Review by Mugurell — COMMENTED** [review]
First pass.
Overall looks good 👍.
Have to look more closely to the snackbar functionality, I see differences in UI to what the snackbar looks for other features.


https://user-images.githubusercontent.com/11428869/221220452-1a44ad43-3f07-4f67-94d2-777825a172c4.mp4

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt
Comment (t-p-white) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/share/ShareDownloadFeature.kt:29]: Agreed. I had originally updated these to [LifecycleAwareFeature] but thought maybe they were kept for a reason - thanks for highlighting

File: android-components/components/feature/contextmenu/src/test/java/mozilla/components/feature/contextmenu/ContextMenuCandidateTest.kt
Comment (t-p-white) [reply android-components/components/feature/contextmenu/src/test/java/mozilla/components/feature/contextmenu/ContextMenuCandidateTest.kt:2055]: SnackbarDelegate moved to Support-Utils

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt
Comment (t-p-white) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/copy/CopyDownloadFeature.kt:80]: Gotcha. Thanks for raising, that's a good point

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/CopyInternetResourceStateReducerTest.kt
Comment (Mugurell) [inline android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/CopyInternetResourceStateReducerTest.kt:18]: nit: (here and elsewhere): While we still have some tests which don't follow the [`GIVEN-WHEN-THEN` naming scheme](https://martinfowler.com/bliki/GivenWhenThen.html) we are trying to have all new tests following it.

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/InternetResourceReducerUtilsTest.kt
Comment (Mugurell) [inline android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/InternetResourceReducerUtilsTest.kt:18]: The `Assert.assertFalse` /  `Assert.assertEquals` should be imported.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (Mugurell) [inline android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:283]: nit: Code refactoring / cleanup could be left for separate tickets / commits.

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (Mugurell) [reply android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:570]: To have it easily available from more places - eg the new `CopyDownloadFeature`, nice! 👍

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt
Comment (Mugurell) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt:39]: If the functionality is now in the same package this can be set to have an `internal` visibility back.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt
Comment (Mugurell) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/temporary/TemporaryDownloadFeature.kt:63]: nit: Leftover from previous implementations: The class is named `TemporaryDownloadFeature`  but the log tag is `TemporaryDownloadMiddleware`.

File: android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/CopyDownloadFeatureTest.kt
Comment (Mugurell) [inline android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/CopyDownloadFeatureTest.kt:79]: nit: new tests should follow the `GIVEN-WHEN-THEN` naming scheme.

File: android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt
Comment (Mugurell) [inline android-components/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/content/Context.kt:196]: Nice 👍

File: android-components/components/support/utils/build.gradle
Comment (Mugurell) [inline android-components/components/support/utils/build.gradle:34]: Not sure why this is needed?

File: fenix/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt
Comment (Mugurell) [inline fenix/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt:498]:
```suggestion
            snackbarDelegate = FenixSnackbarDelegate(binding.dynamicSnackbarContainer),
```

to ensure the snackbar is shown at the indended position.



https://user-images.githubusercontent.com/11428869/222115107-714f7310-5e73-418d-bb9a-874f97eac179.mp4

**Review by Mugurell — APPROVED** [review]
LGTM.
Left some comments about things to improve but overall this is looking and working pretty nicely.
The currently failing CI tasks seem like intermittent issues.

**Comment by Mugurell** [comment]
The `build-components-compose-engine` error
 > Could not resolve all files for configuration ':compose-engine:releaseRuntimeClasspath'.
     Could not find com.google.android.material:material:1.2.1.
      Required by:
          project :compose-engine > project :browser-state > project :support-utils
          project :compose-engine > project :browser-state > project :support-ktx
          
does seem like something we have to solve.
Thinking that the easiest and best for the future would be increase the version? 
I see that `com.google.android.material:material` [is already at version 1.8.0](https://github.com/material-components/material-components-android/releases).

File: android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/CopyDownloadFeatureTest.kt
Comment (t-p-white) [reply android-components/components/feature/downloads/src/test/java/mozilla/components/feature/downloads/temporary/CopyDownloadFeatureTest.kt:79]: 👍 My thoughts on this were as this is derived from the original ShareDownloadFeatureTest it would be easier to review and maintain if they were consistent in structure and naming?

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/InternetResourceReducerUtilsTest.kt
Comment (t-p-white) [reply android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/InternetResourceReducerUtilsTest.kt:18]: Good spot - automatic import 🙃

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/CopyInternetResourceStateReducerTest.kt
Comment (t-p-white) [reply android-components/components/browser/state/src/test/java/mozilla/components/browser/state/reducer/CopyInternetResourceStateReducerTest.kt:18]: Gotcha - I think there's a case for having these follow the existing ShareInternetResourceStateReducerTest? (same as https://github.com/mozilla-mobile/firefox-android/pull/948#discussion_r1121542335)

File: android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt
Comment (t-p-white) [reply android-components/components/feature/contextmenu/src/main/java/mozilla/components/feature/contextmenu/ContextMenuCandidate.kt:283]: Will bear this in mind for future issues 👍

File: android-components/components/support/ktx/build.gradle
Comment (jonalmeida) [inline android-components/components/support/ktx/build.gradle:44]: Drive-by 🚗 : A bit unfortunate to have a base support library need to depend on a UI dependency. It looks like we only need it for `Snackbar.LENGTH_LONG`?

**Comment by t-p-white** [comment]
To land on the 06/03/23

File: android-components/components/support/ktx/build.gradle
Comment (t-p-white) [reply android-components/components/support/ktx/build.gradle:44]: Agreed, I did think about changing SnackbarDelegate but as it already existed avoided making the changes. My thoughts were: 

- document the existing duration param to describe the accepted int vals e.g. -1 for short (open to user error)
- have a duration enum in snackbar delegate (felt like duplication of Snackbar)
- change duration to a boolean e.g true for long, false for short (doesn't seem particularly clean)
- have showShort and showLong funcs (my preference)

Something like this:


```
    override fun showLong(
        snackBarParentView: View,
        text: Int,
        duration: Int,
        action: Int,
        listener: ((v: View) -> Unit)?,
    ) {
        show(snackBarParentView, text, Snackbar.LENGTH_LONG, action, listener)
    }

    override fun showShort(
        snackBarParentView: View,
        text: Int,
        duration: Int,
        action: Int,
        listener: ((v: View) -> Unit)?,
    ) {
       show(snackBarParentView, text, Snackbar.LENGTH_SHORT, action, listener)
    }

    fun show(
        snackBarParentView: View,
        text: Int,
        duration: Int,
        action: Int,
        listener: ((v: View) -> Unit)?,
    ) {
        val snackbar = Snackbar.make(
            snackBarParentView,
            text,
            duration,
        )

        if (action != 0 && listener != null) {
            snackbar.setAction(action, listener)
        }

        snackbar.show()
    }
```

Interested to get your thoughts.

Comment (jonalmeida) [reply android-components/components/support/ktx/build.gradle:44]: The last option makes sense to me (assuming the `duration` param on the override methods are a typo). Where would those two overrides live though? In `feature-contextmenu` I guess (which is fine)?

Comment (t-p-white) [reply android-components/components/support/ktx/build.gradle:44]: @jonalmeida yep typo! My thinking was to move it all within the SnackbarDelegate. I've moved this to another issue to unblock this https://bugzilla.mozilla.org/show_bug.cgi?id=1820467

**Pull request merged**
